### PR TITLE
WP-10259 CSV Import Performance

### PR DESCRIPTION
**Pull request by jayma91**
WP-10259 CSV Import Performance
https://varicent.atlassian.net/browse/WP-10259

notes:
- added dialect detection code for running tap-csv imports
- for tap-csv imports using tap-s3-csv, tap config will not have external_id

### CODE REVIEW

File: tap_s3_csv/dialect.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    # there is only one table in the array
+    for table in config['tables']:
+        # will return all matching files in s3 with given prefix and table name in config
+        s3_files = s3.get_input_files_for_table(config, table)
+
+        for s3_file in s3_files:
+            detect_dialect(config, s3_file, table)
+
+
+def detect_dialect(config, s3_file, table):
```
Comment (jayma91) [inline tap_s3_csv/dialect.py:19]: Followed dialect detection code in tap-csv.

https://github.com/symon-ai/tap-csv/blob/749ac1818611dc1bc55a84ccf4abee0e495c686d/tap_csv/__init__.py#L205

https://github.com/symon-ai/tap-csv/blob/749ac1818611dc1bc55a84ccf4abee0e495c686d/tap_csv/__init__.py#L16

File: tap_s3_csv/__init__.py [hunk 2]
```diff
@@ -99,6 +100,9 @@ def main():
                 break
         except BaseException as err:
             LOGGER.error(err)
+        
+        # If not external source, it is from importing csv (replacement for tap-csv) 
+        dialect.detect_tables_dialect(config)
```
Comment (kchan-varicent) [inline tap_s3_csv/__init__.py:105]: why don't we need to detect dialect if it's external?

File: tap_s3_csv/dialect.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    config_delimiter = table.get('delimiter', '')
+    config_quotechar = table.get('quotechar', '')
+    config_encoding = table.get('encoding', '')
+
+    detect_delimiter = config_delimiter == ''
+    detect_quotechar = config_quotechar == ''
+    detect_encoding = config_encoding == ''
+
+    if not detect_delimiter and not detect_delimiter and not detect_quotechar:
```
Comment (kchan-varicent) [inline tap_s3_csv/dialect.py:28]: typo: missing detect_encoding

File: tap_s3_csv/dialect.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    FIRST_CHARDET_LINES = MAX_CHARDET_LINES / 10
+    interesting = []
+    interesting_map = {}
+
+    lines = []
+    lines_read = 0
+
+    file_key = s3_file.get('key')
+    file_handle = s3.get_file_handle(config, file_key)
+    file_iter = file_handle.iter_lines()
```
Comment (kchan-varicent) [inline tap_s3_csv/dialect.py:65]: does calling this download the entire file? tap-s3-csv seems to read files differently with raw_stream and such.

File: tap_s3_csv/csv_iterator.py [hunk 5]
```diff
@@ -12,9 +12,23 @@ def get_row_iterator(iterable, options=None):
...
+        quotechar=options.get('quotechar', '"'))
 
     headers = set(reader.fieldnames)
+
+    # Check for duplicate columns
+    if len(reader.fieldnames) != len(headers):
+        fieldname_count = {}
+        duplicate_cols = set()
+        for fieldname in reader.fieldnames:
+            fieldname_count[fieldname] = fieldname_count.get(fieldname, 0) + 1
```
Comment (kchan-varicent) [inline tap_s3_csv/csv_iterator.py:25]: why do we care about count? isn't it enough for the if to be `if fieldname in filedname_count:`?

File: tap_s3_csv/csv_iterator.py [hunk 6]
```diff
@@ -12,9 +12,23 @@ def get_row_iterator(iterable, options=None):
...
+            if fieldname_count[fieldname] > 1:
+                duplicate_cols.add(fieldname)
+        ', '.join(duplicate_cols)
```
Comment (kchan-varicent) [inline tap_s3_csv/csv_iterator.py:28]: not setting this to anything.

File: tap_s3_csv/__init__.py (hunk 2 above)
Comment (jayma91) [reply tap_s3_csv/__init__.py:105]: We only have external_id in config for s3 connectors. For s3 connectors, it had a config modal in UI where user could select the delimiter to use. So the original behaviour for s3 connectors was to use the delimiter selected by user along with default quotechar and utf-8 encoding whereas for csv imports, we just upload the csv file and detect dialect in tap. I tried not to change the original behaviour, and also, it was ambiguous which dialect to use when detected delimiter differed from the one passed in by user. Should we still detect dialects for s3 connectors?

File: tap_s3_csv/csv_iterator.py (hunk 5 above)
Comment (jayma91) [reply tap_s3_csv/csv_iterator.py]: fixed.

File: tap_s3_csv/csv_iterator.py (hunk 6 above)
Comment (jayma91) [reply tap_s3_csv/csv_iterator.py]: fixed.

File: tap_s3_csv/__init__.py (hunk 2 above)
Comment (kchan-varicent) [reply tap_s3_csv/__init__.py:105]: hmm no but maybe we should have originally detected. it's fine, leave it the way it is.

**Review by kchan-varicent — APPROVED** [review]

**Comment by jayma91** [comment]
@rsantos-varicent Could you create a tag for this? (v1.3.8)

**Pull request merged**